### PR TITLE
Features/memory

### PR DESCRIPTION
**Pull request by jonnerloller**
Added a bunch of features.
1. Memory Manager
2. Basic Logger (right now it just rediverts it to std::ostream and then couts it, but we can change that later
3. NonPersistantString. Basically a way to non-guiltily allocate strings during a frame. lifespan of the string should be < 1 frame. i.e should be only used as a method of passing arguments, or debug text.
4.MemoryStack
5.MemoryHeap
6.templated New,Delete,Allocate,DeAllocate. for normal things and arrays.
7. Some basic memory stat stuff which i need to remind myself to use.
8. Basic primitive tests case. Used for my Nonpersistantstring.

Note.
Code is not anywhere near perfect. Would recommend leaving a review and accepting the request so that we can do a first merge, and you other features can start using this base, before improvements are made.
Any important stuff can be turned into issues and assigned to me

### CODE REVIEW

**Comment by jonnerloller** [comment]
whoops. didn't "resolve some conflicts, but i gueses it didnt matter anyway"

**Comment by jonnerloller** [comment]
bump. Want to continue soon. Might also want to fool around with DX12. but high level before low level.

**Comment by wagk** [comment]
Currently VS2017 fails compilation with `sizedtypes.h`, something about undefined defines

**Comment by jonnerloller** [comment]
Done the necessary Solution configurations to make it compilable

File: kurobako/src/enginemain.cpp [hunk 1]
```diff
@@ -1,31 +1,31 @@
...
-		++i;
-
-		int*p = nullptr;
-#if defined(KUROBAKODEBUG)
-		*p = 3;
+        constexpr uint64 allocation_size = static_cast<uint64>(2) << 30;
+	    kurobako::memory::MemoryManager::InitializeMemoryManager(allocation_size);
+	    CREATE_SINGLETON_SCOPED(kurobako::log, Logger);
+	
+#if defined(KBK_DEBUG)
```
Comment (wagk) [inline kurobako/src/enginemain.cpp:16]: Are we using `KBK_*` as a define prefix now?

Comment (wagk) [reply kurobako/src/enginemain.cpp:16]: I mean I don't mind but let's fix this while we're here

File: kurobako/src/log/logger.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#define LOG_INFO(str,level) GET_SINGLETON_SCOPED(kurobako::log,Logger)->Log(kurobako::NonPersistentString("%s(%d) [INFO] > %s",__FILE__,__LINE__,str.Get()),level)
+#define LOG_ERROR(str,level) GET_SINGLETON_SCOPED(kurobako::log,Logger)->Log(kurobako::NonPersistentString("%s(%d) [ERROR] > %s",__FILE__,__LINE__,str.Get()),level)
+#define LOG_DEBUG(str,level) GET_SINGLETON_SCOPED(kurobako::log,Logger)->Log(kurobako::NonPersistentString("%s(%d) [DEBUG] > %s",__FILE__,__LINE__,str.Get()),level)
+namespace kurobako::log
+{
+    class Logger
+    {
+		enum
+		{
+			LOG_BUFFER_SIZE = 1 << 20
```
Comment (wagk) [inline kurobako/src/log/logger.h:17]: we probably also need to standardize tabs/spaces, or get a linter

**Review by wagk — APPROVED** [review]

File: kurobako/src/enginemain.cpp (hunk 1 above)
Comment (jonnerloller) [reply kurobako/src/enginemain.cpp:16]: I was using KUROBAKO but it became too long.

Lets fix it as

KBK_MACRO_DEFINE
kbkMacroFunction

File: kurobako/src/log/logger.h (hunk 2 above)
Comment (jonnerloller) [reply kurobako/src/log/logger.h:17]: Spaces, 4.

Will probably use power tools to convert everything to spaces -.- Visual studio keeps resetting my settings for god uinknown reasons

**Comment by jonnerloller** [comment]
Thanks for the pull request.

Probably need to set up the Job Graph. So we can multithread unit tests. makes it test more things too. since we can test the scheduler as well as any multi thread dependant things like lists, or memory, or stuff

**Comment by wagk** [comment]
Keeping the branch (maybe rename it to `feature/memory` for consistency) so further work/refactoring on memory can operate there

File: kurobako/src/log/logger.h (hunk 2 above)
Comment (wagk) [reply kurobako/src/log/logger.h:17]: ok, we'll keep it at 4 spaces, no tabs

**Pull request merged**